### ISSUE TITLE
openjpeg is not working properly

### ISSUE THREAD

**Issue by Redsandro** [internetarchive/archive-pdf-tools#41]
**Recommended actions discussed in this issue:**

- [x] Remove `-threads` (or place flag last) for OpenJpeg (**done:** https://github.com/internetarchive/archive-pdf-tools/commit/31def81e0ddd995a40ec1a7f1de95b5e2375407f)
- [ ] Allow threads to be specified (encoder agnostic e.g. `-num_threads` for Kakadu)
- [ ] Merge debug messages from ROI build

(Original issue below.)

---

Using `-J openjpeg` results in lossless compression. Probably because it is the default:

```bash
$ opj_compress -h | grep -A 3 "Default encoding"
Default encoding options:
-------------------------

 * Lossless
```

My test image compresses 0.29 times (got 3 times bigger).

```bash
recode_pdf -v --dpi 300 \
  -J openjpeg \
  -I in.png --hocr-file in.hocr -o out-openjpeg.pdf
```

`recode_pdf` should probably use some sane defaults with `-J openjpeg`.

### Command line arguments don't work.

Unfortunately, manually setting the compression options doesn't work. According to `opj_compress -h`, the compression ratio can be adjusted with `-q` or `-r`:

``` 
-r <compression ratio>,<compression ratio>,...
    Different compression ratios for successive layers.
    The rate specified for each quality level is the desired
    compression factor (use 1 for lossless)
    Decreasing ratios required.
      Example: -r 20,10,1 means 
            quality layer 1: compress 20x, 
            quality layer 2: compress 10x 
            quality layer 3: compress lossless
    Options -r and -q cannot be used together.
-q <psnr value>,<psnr value>,<psnr value>,...
    Different psnr for successive layers (-q 30,40,50).
    Increasing PSNR values required, except 0 which can
    be used for the last layer to indicate it is lossless.
    Options -r and -q cannot be used together.
```

Yet the resulting files are identical size-wise regardless of compression-flags:

```bash
recode_pdf -v --dpi 300 \
  --fg-compression-flags ' -r 20' \
  --bg-compression-flags ' -r 20' \
  -J openjpeg \
  -I in.png --hocr-file in.hocr -o out-openjpeg.pdf
```

```bash
recode_pdf -v --dpi 300 \
  --fg-compression-flags ' -q 5' \
  --bg-compression-flags ' -q 5' \
  -J openjpeg \
  -I in.png --hocr-file in.hocr -o out-openjpeg.pdf
```

Am I missing something, or is this a bug with either `recode_pdf` or the documentation?

### Testing openjpeg directly

```bash
$ opj_compress -r 750 -i in.png -o out.jp2
```

3.0 MB -> 34,3 kB

### Additional info

Linux Mint 20.2 AKA Ubuntu 20.04.3

**openjpeg version**

```bash
$ opj_compress -h | grep openjp2
It has been compiled against openjp2 library v2.3.1.
```

### Test scan to experiment with

[test_1.png.zip](https://github.com/internetarchive/archive-pdf-tools/files/8104422/test_1.png.zip)

**Comment by MerlijnWajer** [comment]
What version of internetarchivepdf did you use?

**Comment by Redsandro** [comment]
Ah, yes, I just typed `recode_pdf --version` and now I have a deja-vu where I decided to reply once I figured out how to answer that question. :blush: 

> recode_pdf: error: unrecognized arguments: --version

I don't think I remember how/where I built this. Can you give me a hint on how to answer your question?

**Comment by Redsandro** [comment]
This dir is in my `/usr/local/lib/python3.8/dist-packages`: `archive_pdf_tools-1.4.12-py3.8-linux-x86_64.egg`

**Comment by MerlijnWajer** [comment]
Ah, I see. I'll add that flag. I'll test using your input files momentarily.

**Comment by MerlijnWajer** [comment]
I get this:

```
$ recode_pdf -v --dpi 300   -J openjpeg   -I /tmp/test_1.png --hocr-file /tmp/test.hocr -o out-openjpeg.pdf
	 MMX
	 SSE
	 SSE2
	 SSE3
	 SSSE3
	 SSE41
	 POPCNT
	 SSE42
	 AVX
	 F16C
	 FMA3
	 AVX2
Creating text only PDF
Starting page generation at 2022-05-02T17:28:54.211116
Finished page generation at 2022-05-02T17:28:54.276476
Creating text pages took 0.0654 seconds
Inserting (and compressing) images
Converting with image mode: 2
MRC time breakdown: {'image_load': 233, 'grey_conversion': 10, 'hocr_mask_gen': 176, 'est_1': 41, 'threshold': 145, 'fast_denoise': 16, 'mask_jbig2': 90, 'fg_partial_blur': 338, 'fg_jp2': 3068, 'bg_partial_blur': 343, 'bg_jp2': 2330, 'page_image_insertion': 0}
Saving PDF now
Processed 1 pages at 7.00 seconds/page
Compression ratio: 21.708409
```

**Comment by MerlijnWajer** [comment]
[out-openjpeg.pdf](https://github.com/internetarchive/archive-pdf-tools/files/8604963/out-openjpeg.pdf)

**Comment by MerlijnWajer** [comment]
My opj_compress and opj_decompress are linked against openjp2 v2.4.0

**Comment by MerlijnWajer** [comment]
I can however confirm that the fg and bg compression flags don't seem to work.

**Comment by MerlijnWajer** [comment]
Ok, this happens because as soon as opj_compress or opj_decompress find an argument that is just a space, they stop parsing alltogether, without erroring out.

**Comment by MerlijnWajer** [comment]
With a fix that I will push momentarily at least the args seem to work:

```
$ recode_pdf -v --dpi 300 --fg-compression-flags ' -r 500' --bg-compression-flags ' -r 500'   -J openjpeg   -I /tmp/test_1.png --hocr-file /tmp/test.hocr -o out-openjpeg.pdf
	 MMX
	 SSE
	 SSE2
	 SSE3
	 SSSE3
	 SSE41
	 POPCNT
	 SSE42
	 AVX
	 F16C
	 FMA3
	 AVX2
Creating text only PDF
Starting page generation at 2022-05-02T18:04:33.287084
Finished page generation at 2022-05-02T18:04:33.326647
Creating text pages took 0.0396 seconds
Inserting (and compressing) images
Converting with image mode: 2
Call: ['opj_compress', '-i', '/tmp/fgo92jcv5f.pnm', '-o', '/tmp/fgm3nxdfag.jp2', '-threads', '1', '-r', '500']
Call: ['opj_compress', '-i', '/tmp/bgkpyidxh7.pnm', '-o', '/tmp/bg66z2xymx.jp2', '-threads', '1', '-r', '500']
MRC time breakdown: {'image_load': 230, 'grey_conversion': 9, 'hocr_mask_gen': 172, 'est_1': 39, 'threshold': 145, 'fast_denoise': 16, 'mask_jbig2': 110, 'fg_partial_blur': 389, 'fg_jp2': 3422, 'bg_partial_blur': 383, 'bg_jp2': 2393, 'page_image_insertion': 1}
Saving PDF now
Processed 1 pages at 7.48 seconds/page
Compression ratio: 19.327876
```

**Comment by MerlijnWajer** [comment]
Care to install the wheel from here and give it a try? https://github.com/internetarchive/archive-pdf-tools/suites/6346428522/artifacts/228791471

**Comment by Redsandro** [comment]
Thank you @MerlijnWajer for figuring this out.

> Care to install the wheel from here and give it a try? https://github.com/internetarchive/archive-pdf-tools/suites/6346428522/artifacts/228791471

Absolutely. Can you advise me on how to best install a wheel when my current version is installed using `python3 ./setup.py install`?

**Comment by MerlijnWajer** [comment]
I think you need to pick the wheel for your platform from that zip and run `pip install --force-reinstall -U path-to-wheel.whl --no-deps` - that should be all

**Comment by MerlijnWajer** [comment]
I'd recommend to run this from your virtual env, assuming you used one.

**Comment by Redsandro** [comment]
I am now testing `archive_pdf_tools-1.4.13-cp38-cp38-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_12_x86_64.manylinux2010_x86_64.whl`.

I can confirm that the version argument works. It is very convenient. :+1: 

```bash
$ recode_pdf --version
internetarchivepdf 1.4.13
```

Using default args, compression is still.. well.. ridiculous at 0.29.

```bash
$ recode_pdf -v -J openjpeg \
>   -I in.png --hocr-file in.hocr -o out-openjpeg.pdf
	 MMX SSE SSE2 SSE3 SSSE3 SSE41 POPCNT SSE42 AVX F16C FMA3 AVX2
Creating text only PDF
Starting page generation at 2022-05-04T16:29:27.359975
Finished page generation at 2022-05-04T16:29:27.377325
Creating text pages took 0.0174 seconds
Inserting (and compressing) images
Converting with image mode: 2
MRC time breakdown: {'image_load': 112, 'grey_conversion': 7, 'hocr_mask_gen': 115, 'est_1': 27, 'threshold': 100, 'fast_denoise': 14, 'mask_jbig2': 59, 'fg_partial_blur': 210, 'fg_jp2': 2091, 'bg_partial_blur': 219, 'bg_jp2': 1442, 'page_image_insertion': 7}
Saving PDF now
Processed 1 pages at 4.53 seconds/page
Compression ratio: 0.291197
```

If I'm not misreading, you are getting a respectable 21.7 compression using the same arguments and the same test file I provided. So this is puzzling. Because that could tell us the problem is with my version of `openjpeg`, but as originally reported, it compresses fine using manual `opj_compress -r 750 -i in.png -o out.jp2` -> compression ratio of ~90.

Is `recode_pdf` not using `opj_compress` internally? Can you reconstruct the exact command `recode_pdf` uses so I can test it manually?

:bulb: a related helpful addition to the verbose mode (`-v`) would be to write out the commands used internally to compress the image files.

**Comment by MerlijnWajer** [comment]
Let me cut a new release and see if that version solves the problem for you. I can add the verbose mode.

**Comment by MerlijnWajer** [comment]
I will add the verbose mode after the release, as I don't think I want verbose mode to cause it to print debug info, that probably needs to become a different (debug) flag.

**Comment by MerlijnWajer** [comment]
I uploaded version 1.4.14 - can you re-test to make sure it doesn't resolve your problems?

**Comment by Redsandro** [comment]
> I uploaded version 1.4.14 - can you re-test to make sure it doesn't resolve your problems?

Now I'm getting an error. I downloaded from https://github.com/internetarchive/archive-pdf-tools/issues/33#issuecomment-1118582753. Was that the wrong 1.4.14?

```bash
$ recode_pdf -v --dpi 300   -J openjpeg   -I in.png --hocr-file in.hocr -o out-openjpeg.pdf
	 MMX, SSE, SSE2, SSE3, SSSE3, SSE41, POPCNT, SSE42, AVX, F16C, FMA3, AVX2
Creating text only PDF
Starting page generation at 2022-05-05T20:17:10.715108
Finished page generation at 2022-05-05T20:17:10.741184
Creating text pages took 0.0261 seconds
Inserting (and compressing) images
Converting with image mode: 2
['opj_compress', '-i', '/tmp/fgqylok4it.pnm', '-o', '/tmp/fgfjrbc_ta.jp2', '-threads', '1', '-r', '750']
['opj_compress', '-i', '/tmp/bg7ypal8ig.pnm', '-o', '/tmp/bg6dwqk18d.jp2', '-threads', '1', '-r', '500']
Traceback (most recent call last):
  File "/home/redsandro/.local/bin/recode_pdf", line 306, in <module>
    res = recode(args.from_pdf, args.from_imagestack, args.dpi, args.hocr_file,
  File "/home/redsandro/.local/lib/python3.8/site-packages/internetarchivepdf/recode.py", line 668, in recode
    insert_images_mrc(outdoc, hocr_file,
  File "/home/redsandro/.local/lib/python3.8/site-packages/internetarchivepdf/recode.py", line 436, in insert_images_mrc
    mask_f, bg_f, bg_s, fg_f, fg_s = encode_mrc_images(mrc_gen,
  File "/home/redsandro/.local/lib/python3.8/site-packages/internetarchivepdf/mrc.py", line 667, in encode_mrc_images
    remove(mask_img_pgm)
TypeError: remove: path should be string, bytes or os.PathLike, not NoneType
```

Problem is that the repo version (`setup.py install`) ends up in system at `/usr/local/bin` and the wheels end up in user home at `/home/redsandro/.local/lib` and now I can't use the system version (repo version) anymore because the environments conflict. I can only use the wheels until I figure out how to clean this up. :sweat_smile:

**Comment by MerlijnWajer** [comment]
Hm... that error likely comes from the region of interest branch/code. I have fixed this problem here: https://github.com/internetarchive/archive-pdf-tools/actions/runs/2278368269

As far as system wide versus wheels go, I'd suggest using a virtualenv, that likely makes your life a lot easier.

**Comment by Redsandro** [comment]
> I'd suggest using a virtualenv, that likely makes your life a lot easier.

@MerlijnWajer even after reading a couple of documents, I am still unsure how to use `venv`. How would you use it in this specific case i.e. with `setup.py`?

Thanks @mara004 that's what I needed. I learned something today that's probably obvious to everyone else.

1. I didn't know about `--user` so `setup.py` kept complaining until I used `sudo`.
2. `pip3 install .` is the same as `python3 setup.py install --user`

That sound about right?

**Comment by mara004** [comment]
> pip3 install . is the same as python3 setup.py install --user

Yes. Note though that the setuptools command-line interface is deprecated and might evenutally be removed, so we are not supposed to call `setup.py` directly anymore. `pip`, `build`, `twine` and similar standards-based tools should be used instead.

@MerlijnWajer I think `archive-pdf-tools` and `archive-hocr-tools` are missing a `pyproject.toml` file, which causes deprecation warnings on installation as per PEP 517/518.
```toml
# minimal example for pyproject.toml
[build-system]
build-backend = "setuptools.build_meta"
requires = [
    "setuptools>=47",
    # additional setup requirements here
]
```

**Comment by Redsandro** [comment]
It may be obvious to some but I think this would have helped me so let's just add some minimal instructions. :smiley: 

https://github.com/internetarchive/archive-pdf-tools/pull/44

**Comment by Redsandro** [comment]
> Hm... that error likely comes from the region of interest branch/code. I have fixed this problem here: https://github.com/internetarchive/archive-pdf-tools/actions/runs/2278368269

hmm.. I still get `Compression ratio: 0.292151`. I don't know why.

```bash
$ recode_pdf --version
internetarchivepdf 1.4.14

$ recode_pdf -v --dpi 300   -J openjpeg   -I in.png --hocr-file in.hocr -o out-openjpeg-2.pdf
	 MMX SSE SSE2 SSE3 SSSE3 SSE41 POPCNT SSE42 AVX F16C FMA3 AVX2
Creating text only PDF
Starting page generation at 2022-05-06T20:27:44.654962
Finished page generation at 2022-05-06T20:27:44.680911
Creating text pages took 0.0260 seconds
Inserting (and compressing) images
Converting with image mode: 2
['opj_compress', '-i', '/tmp/fgh7x0xzks.pnm', '-o', '/tmp/fgrmyj184k.jp2', '-threads', '1', '-r', '750']
['opj_compress', '-i', '/tmp/bgvrnfn48a.pnm', '-o', '/tmp/bgzx_r6oa0.jp2', '-threads', '1', '-r', '500']
MRC time breakdown: {'image_load': 112, 'grey_conversion': 7, 'hocr_mask_gen': 122, 'est_1': 27, 'threshold': 100, 'fast_denoise': 14, 'mask_jbig2': 99, 'fg_partial_blur': 214, 'fg_jp2': 2110, 'bg_partial_blur': 224, 'bg_jp2': 1449, 'page_image_insertion': 13}
Saving PDF now
Processed 1 pages at 4.64 seconds/page
Compression ratio: 0.292151
```

So I copy the logged command from the new version:

```bash
$ opj_compress -i in.png -o out.jp2 -threads 1 -r 750
Invalid option threads
[WARNING] An invalid option has been ignored

[INFO] tile number 1 / 1
[INFO] Generated outfile out.jp2
encode time: 1616 ms 
```

And check the file size:

```bash
$ file in.png && stat -c%s $_
in.png: PNG image data, 2472 x 3484, 8-bit/color RGB, non-interlaced
3137842
sander@black:temp$ file out.jp2 && stat -c%s $_ 
out.jp2: JPEG 2000 Part 1 (JP2)
4831007
```

The `jp2` file is bigger than the original `png` file. Interestingly, opj_ complains about _"Invalid option threads"_

So I remove option `-threads 1`

```bash
$ opj_compress -i in.png -o out2.jp2 -r 750

[INFO] tile number 1 / 1
[INFO] Generated outfile out2.jp2
encode time: 1694 ms 

$ file out2.jp2 && stat -c%s $_ 
out2.jp2: JPEG 2000 Part 1 (JP2)
34252
```

Success! Looks like for Ubuntu you need to remove the -threads parameter, but I don't know if that causes problems on your operating system.

**Comment by Redsandro** [comment]
Now I can finally see the result, and (at least on my machine) the result is really bad, equal to pillow in the other issue #42.

I recommend `-r 220` as a default, `-r 250` at most.

**Comment by MerlijnWajer** [comment]
I see. So your openjpeg version does not support the `-threads` argument. Hrm... I am not sure how to detect that at runtime, I guess I could check upon start and assume the binary doesn't change halfway through. Or maybe having the threads argument at the end of the flags will cause everything else to work, and have it just ignore the threads variable?

Maybe you could test that, in particular, test if this:

```opj_compress -i in.png -o out.jp2 -r 750 -threads 1```

is the same as:

```opj_compress -i in.png -o out.jp2 -r 750```

and not the same as: 

```opj_compress -i in.png -o out.jp2 -threads 1 -r 750```

**Comment by MerlijnWajer** [comment]
BTW: Pillow relies on OpenJPEG to encode/decode JPEG2000, so I would expect their quality/performance to be identical.

**Comment by Redsandro** [comment]
> Maybe you could test that, in particular, test if this:
> 
> `opj_compress -i in.png -o out.jp2 -r 750 -threads 1`
> 
> is the same as:
> 
> `opj_compress -i in.png -o out.jp2 -r 750`
> 
> and not the same as:
> 
> `opj_compress -i in.png -o out.jp2 -threads 1 -r 750`

Yes, you are correct! :+1: :+1: 

So if `-threads` goes last, it doesn't break the rest of the arguments. That would solve the problem.

Of course removing `-threads` would also solve the problem, but you probably included it for a reason. Out of curiosity, why use `-threads 1`? Wouldn't you rather use as much as possible, which is probably the default?

**Comment by MerlijnWajer** [comment]
The reason I specify the threads is more an archive.org default than necessarily what makes more sense for other users: we try to have one job use one core, so having encoding use one thread is just in line with that policy. It's usually also more efficient on a per-core basis. If you think there's a real need for others to specify the amount of threads, I suppose we could add it.

**Comment by Redsandro** [comment]
I understand the decision makes sense when you digitize/archive documents on a large scale or have machines at archive.org run many tasks in parallel. 

When used on a personal basis however, you scan and process one document (as a collection of pages) at a time. During processing, most time by far is spent waiting on `recode_pdf` to the point I get way too many coffee breaks so yes I would prefer it to use all 16 available threads on my computer. It would go maybe 800% faster. (Although for personal tasks I prefer `kakadu`, but I assume the same choice is made there using ` -num_threads 1` I guess.) So yes I believe there is a real need, one flag that automatically translates to the encoder (openjpeg, kakadu, etc).

I would go one step further and suggest that maximum cores/threads probably makes sense as a default for most users. It's an interesting idea to read custom defaults from a file called `~/.archive-pdf-tools-rc` or similar using `ConfigParser` or something so companies and non-profits can store their favorite defaults tailored to their workflow without having to change any existing scripted use of `archive-pdf-tools`.

But if I can just change it myself manually, that would make me happy too. :smiley: 

By the way, the latest version from `master` doesn't show the internal command in verbose mode anymore, e.g.:

```
['opj_compress', '-i', '/tmp/fgh7x0xzks.pnm', '-o', '/tmp/fgrmyj184k.jp2', '-threads', '1', '-r', '750']
```

This is very valuable information. Can you add it back, or add it using a different flag e.g. `-vvv` in stead of `-v` if you prefer to hide it by default?

**Comment by MerlijnWajer** [comment]
Regarding the threads, one thing to keep in mind that none of the other processing code is threaded - only the JPEG2000 encode/decode is, so you will not very likely see a 800% increase. For realtime applications, I suppose it would be useful to be able to specify a certain number of threads to use for processing (where supported).

The graph below shows the usual breakdown for processing times on the archive.org cluster, which sometimes has slow I/O (hence image_load being the top blue line). The slowest operation after image load is the foreground JPEG2000 generation, but even you could make that 8 times faster, in most cases (depending on the type of documents), this will only save 5-10% of the total processing time or so. Of course, with Grok or OpenJPEG this could look different, as OpenJPEG encodes a lot slower.

![breakdown](https://user-images.githubusercontent.com/517897/167270411-accafa98-dd24-4c0e-85a1-33aade7645fc.png)


Regarding the debug messages, yeah, they are only enabled in the `roi` build right now, I didn't add proper `-vv` flag type support yet.

**Comment by Redsandro** [comment]
> Regarding the threads, one thing to keep in mind that none of the other processing code is threaded - only the JPEG2000 encode/decode is, so you will not very likely see a 800% increase.

I understand. Still, most time is spent waiting on that jp2 as you can see in your graph (the pink + the yellow). Fixing that has the most impact. Eyeballing it looks like it takes half the total load. If a 12 core Xeon can do that half 10 times faster, then you're looking at a 45% speed reduction, which is significant. The NUC on my desk has 6 cores, it would be 40% faster on theoretical average.

I think it's worth making the number of threads either configurable or disable-able (if disabled it will usually default to the number of cores available, effectively making it a toggle between all cores and 1 core).

**Comment by MerlijnWajer** [comment]
Right, but the speedup is not linear, so you end up with less pages processed per cpu time / watt. Instead of 8 threads, you can just run 8 parallel compressions. :-) 

But yeah, we could make a flag for it, but the default I think I'd like to be 1 thread.

**Comment by MerlijnWajer** [comment]
And of course, the compression could also be done on multiple -pages- at once, but that will cause ram usage to go up with the same factor.

**Comment by Redsandro** [comment]
> the speedup is not linear, so you end up with less pages processed per cpu time / watt.

I don't know enough about that. Probably. But like I said that's probably a consideration for a big operation like archive.org. Not for one person with a box of documents who wants to be done as quick as possible. I look at it pragmatically. Artificially limiting the encoder to 1 thread takes nearly 1 real second per image on my current machine:

```bash
$ time kdu_compress -i scan.bmp -o out_k.jp2 -no_weights -rate 0.5 Creversible=no Rweight=16 Rlevels=2 -num_threads 1
Processed using the multi-threaded environment, with
    1 parallel threads of execution

real    0m0,924s
user    0m0,884s
sys     0m0,040s
```

Now the default (without limiting num_threads) is the default for a reason. It takes about 200ms:

```bash
$ time kdu_compress -i scan.bmp -o out_k.jp2 -no_weights -rate 0.5 Creversible=no Rweight=16 Rlevels=2
Processed using the multi-threaded environment, with
    16 parallel threads of execution

real    0m0,212s
user    0m1,507s
sys     0m0,022s
```

That's about one 5th the time on an 8 core machine. With dozens of documents, you're quickly eating off minutes of processing time. Hence the too many coffee breaks. :wink: 

> the default I think I'd like to be 1 thread.

That's okay. I made my argument for you to consider and if this default makes the most sense for you then you should keep it. But it would be nice if users could disable/out-out of the artificial hardware limiter.

**Comment by MerlijnWajer** [comment]
Ok, so like you mentioned at the top, the wishlist is:

* Add a --debug flag to debugs the command execution (amongst other things)
* Allow specifying the number of threads to use

And that should be enough? (of course, the actual ratio's for openjpeg we still want to tweak, but there is a separate issue for this)

**Comment by Redsandro** [comment]
> * Add a --debug flag to debugs the command execution (amongst other things)
> * Allow specifying the number of threads to use
> 
> And that should be enough?

Yes, both would be very helpful! :+1:

**Comment by MerlijnWajer** [comment]
So I have almost finished the code, but here's a fair warning:

```
merlijn@gentoo-x13 ~ $ opj_compress -i /tmp/fgyoa6zlzv.pnm -o /tmp/fgo630wb5d.jp2 -q 26 -threads 1

[INFO] tile number 1 / 1
[INFO] Generated outfile /tmp/fgo630wb5d.jp2
encode time: 2184 ms
merlijn@gentoo-x13 ~ $ opj_compress -i /tmp/fgyoa6zlzv.pnm -o /tmp/fgo630wb5d.jp2 -q 26 -threads 16

[INFO] tile number 1 / 1
[INFO] Generated outfile /tmp/fgo630wb5d.jp2
encode time: 3829 ms
merlijn@gentoo-x13 ~ $ opj_compress -i /tmp/fgyoa6zlzv.pnm -o /tmp/fgo630wb5d.jp2 -q 26 -threads 16

[INFO] tile number 1 / 1
[INFO] Generated outfile /tmp/fgo630wb5d.jp2
encode time: 3802 ms
merlijn@gentoo-x13 ~ $ opj_compress -i /tmp/fgyoa6zlzv.pnm -o /tmp/fgo630wb5d.jp2 -q 26 -threads 16

[INFO] tile number 1 / 1
[INFO] Generated outfile /tmp/fgo630wb5d.jp2
encode time: 3817 ms
merlijn@gentoo-x13 ~ $ opj_compress -i /tmp/fgyoa6zlzv.pnm -o /tmp/fgo630wb5d.jp2 -q 26 -threads 2

[INFO] tile number 1 / 1
[INFO] Generated outfile /tmp/fgo630wb5d.jp2
encode time: 2207 ms
merlijn@gentoo-x13 ~ $ opj_compress -i /tmp/fgyoa6zlzv.pnm -o /tmp/fgo630wb5d.jp2 -q 26 -threads 1

[INFO] tile number 1 / 1
[INFO] Generated outfile /tmp/fgo630wb5d.jp2
encode time: 2202 ms
```


OpenJPEG just seems waste resources with more threads. I'm sure it depends on the size of your images, but I'm not particularly impressed.

**Comment by MerlijnWajer** [comment]
See following commits:

* threads: https://github.com/internetarchive/archive-pdf-tools/commit/73c78fae9c69663c3cb459df05f5140487321bdc
* debug: https://github.com/internetarchive/archive-pdf-tools/commit/71739a35a2d48f002c90bf67a4f08084007ec942

Please let me know how this works for you.

**Comment by MerlijnWajer** [comment]
Of course, with `-J Pillow` this won't currently work.

**Comment by Redsandro** [comment]
This looks good to me. Thank you. I am not sure what you mean with the debug commit message about getting messy, but this is very helpful. I recommend keeping it.

The timings you showed me from threads are not exactly making me happy either. I know there are circumstances where I observed speed improvements in jp2 compression, and from there on I just assumed there would always be speed improvements although I never tried it with openjpeg. (As you recall from this issue, the `-threads` flag doesn't work on the specific version of openjpeg installed on my machine.)

I can say that I see now - and agree with - the necessity to have 1 thread as a default. I like the option to be able to tweak the number of threads though in an encoder-agnostic fashion. The benefits for openjpeg specifically can always change in the future, as the older version has no such command line option so perhaps with openjpeg the threads flag is still new and in development.

It has an unintended helpful side-effect: If I just want to remember what the flag was called for a specific encoder, I just use -J and --debug. It's helpful.
